# Formation Python

* [Présentation des outils](#Présentation-des-outils)
    * [Git/Github](#Git/Github)
    * [Jupiter Notebook](#Jupiter-Notebook)
    * [Python](#Python)
    * [Modules](#Modules)
        * [Module Pandas](#Module-Pandas)
        * [Module Machine Learning](#Module-Machine-Learning)
* [Github/Git](#Github/Git)
    * [Télécharger un projet en ZIP](#Télécharger-un-projet-en-ZIP)
    * [Créer un repository](#Créer-un-repository)
    * [Push](#Push)
    * [Pull](#Pull)
* [Jupyter](#Jupyter)
    * [Markdown](#Markdown)
    * [Commandes UNIX](#Commandes-UNIX)
    * [Export PDF](#Export-PDF)
    * [Export .py](#Export-.py)
* [Python basic](#Python-basic)
    * [Print](#Print)
    * [Variables](#Variables)
    * [If](#If)
    * [Boucle](#Boucle)
      * [For](#For)
      * [While](#While)
    * [Fonctions](#Fonctions)
* [Modules Python](#Modules-Python)
    * [Pandas](#Pandas)
      * [Lire du CSV](#Lire-du-CSV)
      * [Calcule basique entre colonnes](#Calcule-basique-entre-colonnes)
      * [Pivot](#Pivot)
      * [Export Excel](#Export-Excel)
    * [Machine learning](#Machine-learning)


# Présentation des outils

## Git/Github

En 2016, il s’agit du logiciel de gestion de versions le plus populaire qui est utilisé par plus de douze millions de personnes

Commandes principales:

* git init : crée un nouveau dépôt
* git clone : clone un dépôt distant
* git add : ajoute de nouveaux objets blobs dans la base des objets pour chaque fichier modifié depuis le dernier commit. Les objets précédents restent inchangés
* git commit : intègre la somme de contrôle SHA-1 d'un objet tree et les sommes de contrôle des objets commits parents pour créer un nouvel objet commit
* git branch : crée une nouvelle branche de développement
* git merge : fusionne plusieurs branches de développement

## Jupiter Notebook

Jupyter est une application web utilisée pour programmer dans plus de 40 langages de programmation, dont Julia, Python, R, Ruby ou encore Scala1. Jupyter est une évolution du projet IPython. Jupyter permet de réaliser des notebooks, c'est-à-dire des programmes contenant à la fois du texte en markdown et du code en Julia, Python, R... Ces notebooks sont utilisés en data science pour explorer et analyser des données.

## Python

Python est un langage de programmation objet, il est similaire à Perl, Ruby, Scheme, Smalltalk et Tcl.

Le langage Python est placé sous une licence libre et fonctionne sur la plupart des plates-formes informatiques, des supercalculateurs aux ordinateurs centraux4, de Windows à Unix avec notamment GNU/Linux en passant par macOS, ou encore Android, iOS, et aussi avec Java ou encore .NET. Il est conçu pour optimiser la productivité des programmeurs en offrant des outils de haut niveau (plus proche de l'humain que de la machine) et une syntaxe simple à utiliser.

Il est également apprécié par certains pédagogues qui y trouvent un langage où la syntaxe, clairement séparée des mécanismes de bas niveau, permet une initiation aisée aux concepts de base de la programmation.

## Modules

### Module Pandas

Pandas apporte à Python la possibilité de manipuler de grands volumes de données structurées de manière simple et intuitive. Elle propose en particulier des structures de données et des opérations de manipulation de tableaux numériques et de séries temporelles.

Les principales structures de données sont les séries (pour stocker des données selon une dimension - grandeur en fonction d'un index), les DataFrames (pour stocker des données selon 2 dimensions - lignes et colonnes), les Panels (pour représenter des données selon 3 dimensions, les Panels4D ou les DataFrames avec des index hiérarchiques aussi nommés MultiIndex (pour représenter des données selon plus de 3 dimensions - hypercube)

### Module Machine Learning

# Github/Git

## Récupérer un projet a partir de GitHub

### Télécharger un projet en ZIP

Trouvez un projet intéréssant sur GitHub par exemple:

https://github.com/AdrienMereghetti/Environnement_cours

* cliquez sur "Clone or Download"
* puis sur "Download ZIP"
![DL Zip](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_zip/zip.png)

* Déziper le fichier téléchargé
![DL](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_zip/download_file.png)

* Le projet GitHub ce trouve maintenant sur votre poste 
![File](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_zip/into_file.png)

### Cloner un projet

#### Sur Linux/Mac

* Ouvrez un terminal
![1](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_clone/1.png)
* placez-vous dans le repertoire dans lequel vous voulez récupéré le projet:
```
cd ~/dev
```
![2](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_clone/2.png)
* Créez le repertoire du projet
```
mkdir environnement_cours
```
![3](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_clone/3.png)
* Allez sur GitHub puis copier le lien https de clonnage
![4](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_clone/4.png)
* Puis tapez dans votre terminal:
```
git clone https://github.com/AdrienMereghetti/Environnement_cours.git environnement_cours
```
![5](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_clone/5.png)
* Le projet GitHub ce trouve maintenant sur votre poste
![6](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_clone/6.png)

#### Sur Windows

##### Installation de Git

![1](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_clone_windows/1.png)

![2](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_clone_windows/2.png)

![3](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_clone_windows/3.png)

![4](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_clone_windows/4.png)

![5](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_clone_windows/5.png)

![6](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_clone_windows/6.png)

![7](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_clone_windows/7.png)

![8](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_clone_windows/8.png)

![9](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_clone_windows/9.png)

![10](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_clone_windows/10.png)

![11](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_clone_windows/11.png)

![12](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_clone_windows/12.png)

##### Lancement et utilisation de Git

* Ouvrez le terminal "Git Bash"

![13](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_clone_windows/13.png)

![14](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_clone_windows/14.png)

* placez-vous dans le repertoire dans lequel vous voulez récupéré le projet:
```
cd Documents/
```

![15](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_clone_windows/15.png)

* Créez le repertoire du projet
```
mkdir environnement_cours
```

![16](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_clone_windows/16.png)

* Allez sur GitHub puis copier le lien https de clonnage

![4](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_clone/4.png)

* Puis tapez dans votre terminal:
```
git clone https://github.com/AdrienMereghetti/Environnement_cours.git environnement_cours
```

![17](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/git_clone_windows/17.png)

* Le projet GitHub ce trouve maintenant sur votre poste

## Créer un repository

In [1]:
for i in range(1,8+1):
    print("![%s](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/new_repo/%s.png)"%(i,i))

![1](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/new_repo/1.png)
![2](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/new_repo/2.png)
![3](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/new_repo/3.png)
![4](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/new_repo/4.png)
![5](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/new_repo/5.png)
![6](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/new_repo/6.png)
![7](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/new_repo/7.png)
![8](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/new_repo

* Crééez un compte sur GitHub
* Loggez-vous
* Cliquez sur le +
* Puis "New repository"

![1](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/new_repo/1.png)

![2](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/new_repo/2.png)

![3](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/new_repo/3.png)

![4](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/new_repo/4.png)

![5](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/new_repo/5.png)

![6](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/new_repo/6.png)

![7](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/new_repo/7.png)

![8](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/new_repo/8.png)

## Push

## Pull

# Jupyter

## Markdown

Dans Jupyter on as Deux principaux type de bloc:
* les bloc Markdown comme celui-ci utilisé principalement pour documenter un Notebook
* les bloc de code

Ci-contre un lien vers les principales syntaxes du Markdown
https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet

## Commandes UNIX

Dans un bloc de type "code" dans Jupyter, nous pouvons executer des commandes unix en mettant "!" devant celle-ci

In [2]:
!ls -lsah

total 160K
4.0K drwxrwxr-x   5 jovyan  1000 4.0K Sep 27 15:47 .
4.0K drwxr-xr-x 102 root   root  4.0K Sep 27 15:27 ..
 28K -rw-rw-r--   1 jovyan  1000  25K Sep 14 13:06 API Twitter.ipynb
4.0K drwxrwxr-x   2 jovyan  1000 4.0K Sep 26 16:03 csv
108K -rw-r--r--   1 jovyan users 107K Sep 27 15:47 Formation_Python.ipynb
4.0K -rw-rw-r--   1 jovyan  1000   19 May  2 07:59 .gitignore
4.0K drwxrwxr-x   8 jovyan  1000 4.0K Sep 27 15:42 images
4.0K drwxr-xr-x   2 jovyan users 4.0K Sep 14 13:30 .ipynb_checkpoints


Si une library Python est manquante, nous pouvons aussi l'installer via une ligne de commande linux:
```
    !pip install NomDuModuleManquant
```

In [3]:
!pip install pymongo

## Export PDF

Il est possible d'exporter sous un format PDF 

![1](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/Jupyter/1.png)

![2](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/Jupyter/2.png)

## Export .py

Tout un notebook est exportable et directement utilisable en .py (script python pure)

![3](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/Jupyter/3.png)

# Python basic

## Commentaire

In [4]:
# Il suffit de mettre "#" au début d'une ligne pour signifier à python que c'est un commentaire

## Print

In [5]:
print("Hello World")

Hello World


## Variables

Pour déclarer une variable:

```python
nom_de_variable = valeur
```


In [6]:
ma_variable = 42
print(ma_variable) 

42


On peut nommer une variable comme on veux a l'exception de quelques variable **reservées** Python 3:

![8](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/Python/1.png)

Il existe différents type de variables en python

Il n'y a pas besoin d'indiquer en Python de quel type est cette variable, il la reconnait seul.

On peut connaitre le type d'un variable via la fonction Python

```Python
type(ma_variable)
```

### Les nombres entiers

Le type entier se nomme **int** en Python (qui correspond à l'anglais « integer », c'est-à-dire entier). La forme d'un entier est un nombre sans virgule.

In [7]:
variable_int = 42
type(variable_int)

int

### Les nombres flottants
Les flottants sont les nombres à virgule. Ils se nomment **float** en Python (ce qui signifie « flottant » en anglais). 
(Python n'étant pas un language Français les nombres à virgules sont écrits avec des "**.**" )

```Python
12.5
```

In [8]:
variable_float = 152.66
type(variable_float)

float

### Les chaînes de caractères
Ce type de donnée permet de stocker une série de lettres comme par exemple une phrase. Ils se nomment **str** en Python (qui correspond à l'anglais « string »)

entre guillemets 
```Python
"ceci est une chaîne de caractères"
```
entre apostrophes
```Python
'ceci est une chaîne de caractères'
```
entre triples guillemets
```Python
"""ceci est une chaîne de caractères"""

"""
ceci est une chaîne de caractères
que l'on peux aussi écrire sur 
plusieurs lignes
"""

```


In [9]:
print('Cette chaîne de caractères')

Cette chaîne de caractères


In [11]:
chaine = 'J'aime le Python!'

SyntaxError: invalid syntax (<ipython-input-11-281db01e4a31>, line 1)

Cette chaîne de caractères ci dessus retourne une erreur car il y à un apostrophes.
Il faut échapper les apostrophes se trouvant au cœur de la chaîne avec "\", il n'apparaitra pas dans le print

In [12]:
chaine = 'J\'aime le Python!'
print(chaine)

J'aime le Python!


In [13]:
# On peux aussi utiliser les quillemet pour utiliser des apostrophe dans la chaine sans problèmes

chaine = "J'aime le Python!"
print(chaine)

J'aime le Python!


### Utilisation des variables

In [14]:
age = 21

In [15]:
# On peut faire toutes sortes de calcules avec les variables numérique (Int ou Float)

age_x2 = age * 2
print(age_x2)

42


In [16]:
# La variable "age" ne change pas
print(age)

21


In [17]:
# On peux incrémenter une variable de deux manière différentes

age = age + 1
print(age)

22


In [18]:
# La syntaxe ci-dessus est claire et intuitive mais assez longue, et les programmeurs, tout le monde le sait, sont des fainéants nés. On a donc trouvé plus court.
age += 1
print(age)

23


In [19]:
# Combiner plusieurs types de variable est possible 

variable_int_float = variable_int * variable_float
print(variable_int_float)

6411.72


In [20]:
# On peux aussi combiner un string avec d'autre type de variable de cette manière:

var_string = "Mon age est:"
var_int = 25
var_mix = "%s %i ans, dans deux ans j'en aurai : %i" % (var_string, var_int, var_int+2)
print(var_mix)

Mon age est: 25 ans, dans deux ans j'en aurai : 27


* **%s** signifie que l'on va insérer un **String**
* **%i** signifie que l'on va insérer un **Integer**
* **%f** signifie que l'on va insérer un **Float**

Ils sont mis entre () dans l'ordre de notre string

## If

In [21]:
annee = 2017

# Si annee est égal à 2017, alors on execute ce qui se trouve dans le "if"
if annee == 2017:
    print('La variable "annee" est bien %i'%annee)
# Sinon on execute ce qui se trouve dans le "else"
else:
    print("annee est incorecte, elle devrait etre 2017 mais elle est égal à %i"%annee)

La variable "annee" est bien 2017


In [22]:
annee = 2042

# Si annee est égal à 2017, alors on execute ce qui se trouve dans le "if"
if annee == 2017:
    print('La variable "annee" est bien %i'%annee)
# Sinon on execute ce qui se trouve dans le "else"
else:
    print("annee est incorecte, elle devrait etre 2017 mais elle est égal à %i"%annee)

annee est incorecte, elle devrait etre 2017 mais elle est égal à 2042


## Boucle

### For

Lorsque l’on souhaite répéter un nombre donné de fois la même instruction ou le même bloc d’instructions, la commande **for** est la plus appropriée.

Admettons que l’on veuille afficher 666 fois la phrase "Metal music rocks !". Voici ce que l’on peut faire.

In [23]:
for i in range(666):   # pour i allant de 0 à 665
    print("Metal music rocks !")

Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !


Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !
Metal music rocks !


In [24]:
for i in range(1, 10):   # pour i allant de 1 à 9 
    print(i**2)
    
# i**2 signifie i au carré 
# i**3 signifie i au cube
# ...
# i**n signigne i exposant n

1
4
9
16
25
36
49
64
81


### While

Le principe de la boucle while est d’exécuter un bloc d’instructions tant que (while in english) une condition donnée est vraie. 

In [25]:
i = 1
# Tant que i est inférieur ou égal à 10 on execute les instructions contenue dans le "while"
while i <= 10:
    print(i)
    # On incrémente i 
    i += 1
    
print('Fini !')

1
2
3
4
5
6
7
8
9
10
Fini !


## Fonctions

Une fonction (ou function) est une suite d'instructions que l'on peut appeler avec un nom.

Les variable utilisées à l'interrieur d'un fonction n'existe que dans cette fonction.

In [26]:
# def signifie que l'on déclare une fonction Python
def donne_moi_LA_reponse():
    # Le return contient ce que va retourner la fonction, ici on ne return qu'un int
    return 42;

donne_moi_LA_reponse()

42

In [27]:
def donne_moi_LA_reponse2():
    a = 21
    b = 1
    b += 1
    a = a * b
    return a

donne_moi_LA_reponse2()

42

On peut aussi definir un argument à une fonction

In [28]:
def table(nb): # nb est un argument obligatoire.
    i = 0
    while i < 10: # Tant que i est strictement inférieure à 10,
        # le nb utilisé ici est celui que l'on a défini en argument
        print(i + 1, "*", nb, "=", (i + 1) * nb)
        i += 1 # On incrémente i de 1 à chaque tour de boucle.
        
table(5)
table(9)
table(11)

1 * 5 = 5
2 * 5 = 10
3 * 5 = 15
4 * 5 = 20
5 * 5 = 25
6 * 5 = 30
7 * 5 = 35
8 * 5 = 40
9 * 5 = 45
10 * 5 = 50
1 * 9 = 9
2 * 9 = 18
3 * 9 = 27
4 * 9 = 36
5 * 9 = 45
6 * 9 = 54
7 * 9 = 63
8 * 9 = 72
9 * 9 = 81
10 * 9 = 90
1 * 11 = 11
2 * 11 = 22
3 * 11 = 33
4 * 11 = 44
5 * 11 = 55
6 * 11 = 66
7 * 11 = 77
8 * 11 = 88
9 * 11 = 99
10 * 11 = 110


# Modules Python

## Pandas

In [29]:
# Pandas étant un module, il nous faut l'importer
import pandas

### Lire du CSV

In [30]:
# df1 = pandas.read_csv("csv/movies.csv")
# df2 = pandas.read_csv("csv/ratings.csv")
# df3 = pandas.read_csv("csv/tags.csv")

# df4 = df1.merge(df2, on="movieId")
# df4 = df4.merge(df3, on=["movieId", "userId"])
# df4=df4.drop(["timestamp_y", "timestamp_x", "userId", "tag"], 1).drop_duplicates()
# df4.to_csv("csv/movies_ratings.csv", index=False)

# df = pandas.read_csv("csv/movies_ratings.csv")

# #On affiche les n première ligne de notre dataframe
# df = df.groupby(["movieId", "title", "genres"])["rating"].mean().reset_index()
# df1 = df[["movieId", "genres"]].copy()

# df.columns

# df1=pandas.concat([pandas.Series(row['movieId'], row['genres'].split('|'))              
#                     for _, row in df1.iterrows()]).reset_index().rename(columns={'index': 'genre', 0: 'movieId'})

# df = df.merge(df1, on="movieId").drop("genres", 1)
# df.to_csv("csv/movies_ratings.csv", index=False)

In [31]:
df = pandas.read_csv("csv/movies_ratings.csv")

In [32]:
#On affiche les 10 première ligne de notre dataframe df

df.head(10)

movieId                               title  rating      genre
0        1                    Toy Story (1995)    5.00  Adventure
1        1                    Toy Story (1995)    5.00  Animation
2        1                    Toy Story (1995)    5.00   Children
3        1                    Toy Story (1995)    5.00     Comedy
4        1                    Toy Story (1995)    5.00    Fantasy
5        5  Father of the Bride Part II (1995)    4.50     Comedy
6       47         Seven (a.k.a. Se7en) (1995)    2.75    Mystery
7       47         Seven (a.k.a. Se7en) (1995)    2.75   Thriller
8      104                Happy Gilmore (1996)    2.00     Comedy
9      111                  Taxi Driver (1976)    5.00      Crime

Si on veut trier sur la colonne "genre"

In [33]:
df.loc[df.genre == "Comedy"]

movieId                                              title  rating  \
3           1                                   Toy Story (1995)    5.00   
5           5                 Father of the Bride Part II (1995)    4.50   
8         104                               Happy Gilmore (1996)    2.00   
15        186                                 Nine Months (1995)    2.50   
20        236                                 French Kiss (1995)    3.00   
32        296                                Pulp Fiction (1994)    4.50   
39        339                     While You Were Sleeping (1995)    2.50   
44        357                 Four Weddings and a Funeral (1994)    2.00   
49        520                   Robin Hood: Men in Tights (1993)    3.00   
54        587                                       Ghost (1990)    4.00   
59        608                                       Fargo (1996)    3.25   
76        745             Wallace & Gromit: A Close Shave (1995)    4.50   
77        778                               Trainspotting (1996)    5.00   
87       1028                                Mary Poppins (1964)    5.00   
96       1079                        Fish Called Wanda, A (1988)    4.50   
98       1080                Monty Python's Life of Brian (1979)    5.00   
122      1256                                   Duck Soup (1933)    5.00   
128      1265                               Groundhog Day (1993)    5.00   
139      1376               Star Trek IV: The Voyage Home (1986)    1.00   
143      1517  Austin Powers: International Man of Mystery (1...    2.00   
145      1639                                 Chasing Amy (1997)    5.00   
153      1732                           Big Lebowski, The (1998)    4.00   
163      2004                   Gremlins 2: The New Batch (1990)    0.50   
166      2012                 Back to the Future Part III (1990)    3.00   
173      2082                           Mighty Ducks, The (1992)    5.00   
177      2174                                 Beetlejuice (1988)    3.50   
183      2359       Waking Ned Devine (a.k.a. Waking Ned) (1998)    3.50   
184      2361                              Pink Flamingos (1972)    4.00   
185      2369                   Desperately Seeking Susan (1985)    3.00   
188      2396                         Shakespeare in Love (1998)    5.00   
...       ...                                                ...     ...   
780     66934              Dr. Horrible's Sing-Along Blog (2008)    4.00   
784     67087                             I Love You, Man (2009)    4.00   
791     67997                                 In the Loop (2009)    5.00   
821     70293                               Julie & Julia (2009)    4.00   
829     71464                              Serious Man, A (2009)    3.50   
839     72226                           Fantastic Mr. Fox (2009)    4.00   
856     76251                                    Kick-Ass (2010)    4.00   
857     77455                  Exit Through the Gift Shop (2010)    5.00   
861     78209                        Get Him to the Greek (2010)    2.00   
867     78499                                 Toy Story 3 (2010)    4.75   
881     79242                     Kids Are All Right, The (2010)    4.50   
905     86833                                 Bridesmaids (2011)    4.00   
917     88235                                  Guard, The (2011)    4.00   
935     91077                            Descendants, The (2011)    4.50   
940     94777       Men in Black III (M.III.B.) (M.I.B.³) (2012)    4.50   
945     94959                            Moonrise Kingdom (2012)    3.50   
948     95135                        Your Sister's Sister (2011)    4.50   
954     96588                               Pitch Perfect (2012)    4.00   
966     97921                     Silver Linings Playbook (2012)    4.00   
993    103372                                   Heat, The (2013)    4.00   
1017   106696                                      Fro

On veut afficher tout les films ayant une note supérieur ou égal à 4

In [34]:
df_bons_films = df.loc[df.rating >= 4]
df_bons_films

movieId                                              title  rating  \
0           1                                   Toy Story (1995)    5.00   
1           1                                   Toy Story (1995)    5.00   
2           1                                   Toy Story (1995)    5.00   
3           1                                   Toy Story (1995)    5.00   
4           1                                   Toy Story (1995)    5.00   
5           5                 Father of the Bride Part II (1995)    4.50   
9         111                                 Taxi Driver (1976)    5.00   
10        111                                 Taxi Driver (1976)    5.00   
11        111                                 Taxi Driver (1976)    5.00   
17        215                              Before Sunrise (1995)    5.00   
18        215                              Before Sunrise (1995)    5.00   
22        246                                 Hoop Dreams (1994)    4.00   
23        247                          Heavenly Creatures (1994)    5.00   
24        247                          Heavenly Creatures (1994)    5.00   
25        260          Star Wars: Episode IV - A New Hope (1977)    4.50   
26        260          Star Wars: Episode IV - A New Hope (1977)    4.50   
27        260          Star Wars: Episode IV - A New Hope (1977)    4.50   
32        296                                Pulp Fiction (1994)    4.50   
33        296                                Pulp Fiction (1994)    4.50   
34        296                                Pulp Fiction (1994)    4.50   
35        296                                Pulp Fiction (1994)    4.50   
36        318                   Shawshank Redemption, The (1994)    4.75   
37        318                   Shawshank Redemption, The (1994)    4.75   
38        337                 What's Eating Gilbert Grape (1993)    5.00   
50        529                 Searching for Bobby Fischer (1993)    4.00   
54        587                                       Ghost (1990)    4.00   
55        587                                       Ghost (1990)    4.00   
56        587                                       Ghost (1990)    4.00   
57        587                                       Ghost (1990)    4.00   
58        587                                       Ghost (1990)    4.00   
...       ...                                                ...     ...   
1082   122886  Star Wars: Episode VII - The Force Awakens (2015)    4.50   
1083   122886  Star Wars: Episode VII - The Force Awakens (2015)    4.50   
1084   122886  Star Wars: Episode VII - The Force Awakens (2015)    4.50   
1085   127114                         The End of the Tour (2015)    4.00   
1086   128360                           The Hateful Eight (2015)    4.50   
1087   128606                                    45 Years (2015)    4.50   
1094   133645                                       Carol (2015)    4.00   
1095   133645                                       Carol (2015)    4.00   
1096   134130                                 The Martian (2015)    4.50   
1097   134130                                 The Martian (2015)    4.50   
1098   134130                                 The Martian (2015)    4.50   
1099   134853                                  Inside Out (2015)    4.50   
1100   134853                                  Inside Out (2015)    4.50   
1101   134853                                  Inside Out (2015)    4.50   
1102   134853                                  Inside Out (2015)    4.50   
1103   134853                                  Inside Out (2015)    4.50   
1104   134853                                  Inside Out (2015)    4.50   
1105   134881                                Love & Mercy (2014)    4.00   
1106   135518                                   Self/less (2015)    4.00   
1107   135518                                   Self/less (2015)    4.00   
1108   135518                                   Self/l

### Export Excel

In [35]:
# On exporte notre selection de film ayant obtenu une note supérieur ou égal à 4 sous un format Excel

df_bons_films.to_excel("csv/best_films.xlsx")

![2](https://raw.githubusercontent.com/AdrienMereghetti/Environnement_cours/development/data/notebooks/images/Python/2.png)

### Calcule basique entre colonnes

In [36]:
# Creation d'un DataFrame simple pour l'exemple
df_a = pandas.DataFrame({
    "a": [1,2,3,4,5,6,7,8,9,10],
    "b": [10,20,30,40,50,60,70,80,90,100]
})

df_a

a    b
0   1   10
1   2   20
2   3   30
3   4   40
4   5   50
5   6   60
6   7   70
7   8   80
8   9   90
9  10  100

In [37]:
# La colonne c sera égal à la colonne a * b
df_a["c"] = df_a["a"] * df_a["b"]
df_a

a    b     c
0   1   10    10
1   2   20    40
2   3   30    90
3   4   40   160
4   5   50   250
5   6   60   360
6   7   70   490
7   8   80   640
8   9   90   810
9  10  100  1000

## Machine learning

## Module scikit-learn 

#### Chargement de la base Iris

![Iris](https://makina-corpus.com/blog/metier/2017/imlp/iris)

Cette base contient des Iris qu'un botaniste, Ronald Fisher, a classés en 1936 à l'aide d'une clef d'identification des plantes (type de pétales, sépale, type des feuilles, forme des feuilles, ...).
Puis, pour chaque fleur classée il a mesuré les longueurs et largeurs des sépales et pétales.

In [1]:
from sklearn import datasets
iris = datasets.load_iris()

In [4]:
# Les noms des paramètres de nos donénes/enregistrements
iris.feature_names

['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

In [3]:
# Affichage des 5 premières lignes
iris.data[:5]

array([[ 5.1,  3.5,  1.4,  0.2],
       [ 4.9,  3. ,  1.4,  0.2],
       [ 4.7,  3.2,  1.3,  0.2],
       [ 4.6,  3.1,  1.5,  0.2],
       [ 5. ,  3.6,  1.4,  0.2]])

In [5]:
# La liste des espèces connues, nos labels de classification, est contenue dans l'attribut target_names
iris.target_names

array(['setosa', 'versicolor', 'virginica'], 
      dtype='<U10')

In [6]:
# Création d'une variable target pour un accès plus facile à cet attribut:
target = iris.target # Les labels associés à chaque enregistrement
target 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [ ]:
# Les targets sont un tableau indiquant le numéro de l'espèce de chaque enregistrement:

In [7]:
for i in [0,1,2]:
    print("classe : %s, nb exemplaires: %s" % (i, len(target[ target == i]) ) )

classe : 0, nb exemplaires: 50
classe : 1, nb exemplaires: 50
classe : 2, nb exemplaires: 50


En résumé, l'échantillon de fleurs propose plusieurs informations:

* Les noms des données disponibles : feature_names

* Les mesures réalisées sur l'échantillon de fleurs connues et déjà classées : data

* Il s'agit de nos informations, des paramètres de nos vecteurs pour chaque fleur

* Le nom de chaque espèce : target_names

* Le classement de chaque enregistrement data dans son espèce: target

Il s'agit de la classe de chaque fleur/vecteur

In [8]:
print(iris.DESCR)

Iris Plants Database

Notes
-----
Data Set Characteristics:
    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20  0.76     0.9565  (high!)
    ============== ==== ==== ======= ===== ====================

    :Missing Attribute Values: None
  

In [ ]:
data = iris.data # Pour un accès plus rapide

Tableau numpy de 2 dimensions de 150 enregistrements de 4 valeurs

In [10]:
type(data), data.ndim, data.shape

(numpy.ndarray, 2, (150, 4))

# ANACONDA